### PR TITLE
Investigate GPU non-determinism

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Closes https://github.com/RasaHQ/research/issues/269

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by dakshvar22** [comment]
/modeltest

```yml
include:
 - dataset: ["all"]
   config: ["all"]
```

**Comment by dakshvar22** [comment]
@tttthomasssss @hsm207 The raw data added to the docs comes from [this datadog dashboard](https://p.datadoghq.eu/sb/j2o619n49tr70kxs-363c33e9bf1d39fbca5dce38b6eb07d5) that i setup as part of the experiments. The aggregates for standard deviation were computed in [this notebook](https://github.com/RasaHQ/research-experiments/blob/non-det/results/non-det-experiments/stdev_analysis.ipynb). Lmk if you have any questions.

File: docs/docs/tuning-your-model.mdx [hunk 1]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
 `tf.nn.sparse_softmax_cross_entropy_with_logits`,
 and `tf.math.unsorted_segment` ops. For more information see [here](https://github.com/tensorflow/community/blob/master/rfcs/20210119-determinism.md)
 
-You can see which featurizers are sparse [here](./components.mdx#featurizers),
+For the above reasons, the models are also not guaranteed to yield the exact same
+performance when trained on a GPU across multiple runs. This even applies to situations where training is run
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:438]: alternatively `when trained on GPU`?

File: docs/docs/tuning-your-model.mdx [hunk 2]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+multiple times on the same training data with the same config
+and random seeds set appropriately while being evaluated on a standard held-out test set.
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:440]: I _think_ there is a comma (or possibly more than one) missing 😅 . At least I think after `appropriately`.

File: docs/docs/tuning-your-model.mdx [hunk 3]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+Internal experiments have shown the following fluctuation in model's performance when trained and evaluated on a held-out
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:441]: `model performance`

File: docs/docs/tuning-your-model.mdx [hunk 4]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+test set of a variety of datasets (experiments were run 5 times on each dataset):
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:442]: `on a variety` instead of `of a variety`.

File: docs/docs/tuning-your-model.mdx [hunk 5]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+
+```
++-----------------------+------------+------------------+------------------+-------------------+
+| Task,                 | Number of  | Average standard | Minimum standard | Maximum standard  |
+| Metric                | datasets   | deviation        | deviation        | deviation         |
++-----------------------+------------+------------------+------------------+-------------------+
+| Intent Classification | 11         | 0.0042           | 2.4e-5           | 0.0176            |
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:449]: Can you give the average here as well? I assume this means sth like `0.87 (+/- 0.042)` and not `87 (+/- 0.042)`. I think making that obvious would be good.

File: docs/docs/tuning-your-model.mdx [hunk 6]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
++-----------------------+------------+------------------+------------------+-------------------+
+| Action Selection      | 5          | 0.0025           | 0.0010           | 0.0053            |
+| Macro F1              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+| Conversation Success  | 5          | 0.0077           | 0.0052           | 0.0103            |
+| Accuracy              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+```
+
+You can expect similar fluctuations in the model's performance when you evaluate them on your dataset.
```
Comment (tttthomasssss) [inline docs/docs/tuning-your-model.mdx:466]: `in model performance`

File: docs/docs/tuning-your-model.mdx (hunk 5 above)
Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:449]: The average standard deviation is across 11 (or the corresponding number of datasets available for the task) datasets, where the actual F1 score would be different for each dataset. So you can't really write something like `0.87 (+/- 0.0042)`

I could mention the range of the metric in the `Task, Metric` column itself, if you think that's helpful.

Comment (tttthomasssss) [reply docs/docs/tuning-your-model.mdx:449]: I think that will help, yes.

File: docs/docs/tuning-your-model.mdx [hunk 7]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+For the above reasons, the models are also not guaranteed to yield the exact same
+performance when trained on a GPU across multiple runs. This even applies to situations where training is run
+multiple times on the same training data with the same config
+and random seeds set appropriately while being evaluated on a standard held-out test set.
+Internal experiments have shown the following fluctuation in model's performance when trained and evaluated on a held-out
+test set of a variety of datasets (experiments were run 5 times on each dataset):
+
+```
++-----------------------+------------+------------------+------------------+-------------------+
+| Task,                 | Number of  | Average standard | Minimum standard | Maximum standard  |
```
Comment (hsm207) [inline docs/docs/tuning-your-model.mdx:446]: Can we mention which GPU we used to run this experiment in case people want to replicate this behavior?

File: docs/docs/tuning-your-model.mdx [hunk 8]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+| Intent Classification | 11         | 0.0042           | 2.4e-5           | 0.0176            |
+| Macro F1              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+| Entity Recognition    | 7          | 0.0019           | 0.0007           | 0.0044            |
+| Macro F1              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+| Response Selection    | 2          | 0.0098           | 0.0003           | 0.0231            |
+| Macro F1              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+| Action Selection      | 5          | 0.0025           | 0.0010           | 0.0053            |
```
Comment (hsm207) [inline docs/docs/tuning-your-model.mdx:458]: Is `Action Selection` referring to predictions from TED Policy?

File: docs/docs/tuning-your-model.mdx [hunk 9]
```diff
@@ -434,9 +434,48 @@ underlying tensorflow ops `tf.sparse.sparse_dense_matmul`,
...
+| Macro F1              |            |                  |                  |                   |
++-----------------------+------------+------------------+------------------+-------------------+
+| Conversation Success  | 5          | 0.0077           | 0.0052           | 0.0103            |
```
Comment (hsm207) [inline docs/docs/tuning-your-model.mdx:461]: What is a "Conversation Success" task?

**Review by hsm207 — CHANGES_REQUESTED** [review]
I have some questions

File: docs/docs/tuning-your-model.mdx (hunk 5 above)
Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:449]: Added

File: docs/docs/tuning-your-model.mdx (hunk 7 above)
Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:446]: We don't mention the datasets here on which this study was done because some of them are private customer datasets, so they can't really reproduce the results.

Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:446]: I've added the information about the GPU as well though

File: docs/docs/tuning-your-model.mdx (hunk 8 above)
Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:458]: The task refers to predictions from the complete ensemble of policies but `TEDPolicy` is the only ML based policy which can result in fluctuations in results.

File: docs/docs/tuning-your-model.mdx (hunk 9 above)
Comment (dakshvar22) [reply docs/docs/tuning-your-model.mdx:461]: Refers to the number of test stories successfully predicted (https://rasa.com/docs/rasa/testing-your-assistant#writing-test-stories)

**Review by hsm207 — APPROVED** [review]
